### PR TITLE
Add challenge: GRPO Surrogate Loss

### PR DESCRIPTION
**Pull request by basilwong**
## Summary

  Adds **easy challenge 109: GRPO Surrogate Loss**.

  The challenge implements the core Group Relative Policy Optimization (GRPO) objective: normalize rewards within each prompt’s sampled group, broadcast the resulting sequence
  advantage across response tokens, apply a clipped policy-ratio objective, add a non-negative reference-policy KL penalty, and reduce to one scalar loss.

## Scope

  - Inputs:
    - `rewards`: `[B, G]`
    - `log_pi`, `log_pi_old`, `log_ref`: `[B, G, S]`
  - Rewards are normalized independently for each prompt group using the population standard deviation.
  - Each group-level advantage is broadcast over its `S` response-token positions.
  - The loss uses PPO-style clipping and the non-negative `k3` reference-policy KL sample estimator.
  - Padding and token masks are intentionally out of scope: every response-token position participates in the reduction.
  - The challenge is classified as **easy** because it is a direct tensor implementation of a fully specified loss.

## Documentation and validation

  - Documents tensor shapes, reduction dimensions, numerical bounds, and fixed-input gradient semantics.
  - Reflows long equations and examples into multi-line aligned MathJax blocks for readability.
  - Adds coverage for:
    - a single-token response (`S = 1`);
    - the maximum supported group size;
    - finite, non-negative `k3` KL penalties for nonzero log-ratio differences.

Modal runtime validation notebook: [GRPO Surrogate Loss notebook](https://modal.com/notebooks/basilwong/main/nb-0E8NPExb11cMk7jd2Pi1t8)

HTML Online Viewer: 
image

## Reference

  [DeepSeekMath: Pushing the Limits of Mathematical Reasoning in Open Language Models — Shao et al. (2024), introducing Group Relative Policy Optimization](https://arxiv.org/abs/2402.03300)

### CODE REVIEW

**Review by shxjames — DISMISSED** [review]

File: challenges/medium/109_grpo_surrogate_loss/challenge.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+from typing import Any, Dict, List
+
+import torch
+from core.challenge_base import ChallengeBase, OutTensor, RandnTensor
+
+
+class Challenge(ChallengeBase):
+    name = "GRPO Surrogate Loss"
+    atol = 0.001
+    rtol = 0.001
```
Comment (kunal-mansukhani) [inline challenges/medium/109_grpo_surrogate_loss/challenge.py]: Are these fine? @shxjames

**Review by kunal-mansukhani — DISMISSED** [review]

File: challenges/medium/109_grpo_surrogate_loss/challenge.py (hunk 1 above)
Comment (shxjames) [reply challenges/medium/109_grpo_surrogate_loss/challenge.py]: Oh oops forgot about that, will fix

**Review by kunal-mansukhani — APPROVED** [review]
